### PR TITLE
Add security conformance test suite for Agent Manager

### PR DESCRIPTION
**Pull request by vish-mv**
## Purpose

Introduce automated security testing for authentication, authorization, roles, AgentID, observability, publisher access, and deployed-agent isolation.

No linked issue.

## Goals

- Detect security regressions early.
- Test security against a running Agent Manager deployment.
- Include security checks in E2E, nightly, and release workflows.

## Approach

Added static security checks and live security suites that reuse the existing E2E deployment. Results and logs are published separately in GitHub Actions.

## User stories

As a developer or release engineer, I want automated security checks so that security regressions are detected before release.

## Release note

Added automated security conformance testing for Agent Manager.

## Documentation

N/A — this change only adds internal test infrastructure.

## Training

N/A

## Certification

N/A — no impact on certification content.

## Marketing

N/A

## Automation tests

- Unit tests
  - Added 6 static security invariant tests.
- Integration tests
  - Added 173 live security tests.
  - All 179 checks pass locally.

## Security checks

- Followed secure coding standards? **Yes**
- Ran FindSecurityBugs plugin and verified report? **N/A — not applicable to these Go tests**
- Confirmed that this PR contains no keys, passwords, tokens, usernames, or other secrets? **Yes**

## Samples

Added a small security probe agent used to test AgentID permissions and runtime isolation.

## Related PRs

N/A

## Migrations

N/A — no migration is required.

## Test environment

- macOS
- Docker and local k3d/Kubernetes deployment
- Go test and Ginkgo
- Local Agent Manager quick-start environment

## Learning

Reused the existing E2E architecture and followed security testing patterns for RBAC, token validation, credential lifecycle, and workload isolation.

## Summary by CodeRabbit

* **New Features**
  * Added comprehensive security validation for authorization scopes, roles, tokens, observability, publisher access, runtime hardening, and identity credentials.
  * Added a security probe fixture for runtime, network, identity, and MCP authorization checks.
  * Added Makefile targets for static and live security testing.

* **Bug Fixes**
  * Improved token handling for scoped and intentionally unscoped requests.

* **Chores**
  * Extended CI workflows with security tests, reports, logs, and longer execution timeouts.
  * Added configuration examples and security testing documentation.

### CODE REVIEW

File: agent-manager-service/api/route_authz_invariant_test.go [hunk 1]
```diff
@@ -0,0 +1,878 @@
...
+	"GET /llm-providers/{providerId}",
+	"GET /llm-proxies/{proxyId}",
+	"GET /mcp-proxies/{proxyId}",
+}
+
+func init() {
+	for _, pattern := range gatewayInternalRoutes {
+		unguardedRouteAllowlist[routeKey{"RegisterGatewayInternalRoutes", "HandleFuncWithValidation", pattern}] =
+			"gateway-to-control-plane endpoint on the separately exposed internal listener."
+	}
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply agent-manager-service/api/route_authz_invariant_test.go:117]: Updated in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: agent-manager-service/api/route_authz_invariant_test.go [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+func thunderAMPScopeCatalog(t *testing.T) map[string]bool {
+	t.Helper()
+
+	const valuesPath = "../../deployments/helm-charts/wso2-amp-thunder-extension/values.yaml"
+	f, err := os.Open(valuesPath)
+	if err != nil {
+		t.Fatalf("open Thunder values: %v", err)
+	}
+	defer f.Close()
+
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply agent-manager-service/api/route_authz_invariant_test.go]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: agent-manager-service/api/route_authz_invariant_test.go [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+func forEachFile(t *testing.T, dir string, fn func(*token.FileSet, *ast.File)) {
+	t.Helper()
+
+	fset := token.NewFileSet()
+	pkgs, err := parser.ParseDir(fset, dir, func(info fs.FileInfo) bool {
+		return !strings.HasSuffix(info.Name(), "_test.go")
+	}, 0)
+	if err != nil {
+		t.Fatalf("parse %s: %v", dir, err)
+	}
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply agent-manager-service/api/route_authz_invariant_test.go:815]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: test/e2e/.env.example [hunk 4]
```diff
@@ -11,6 +11,25 @@ IDP_TOKEN_URL=http://thunder.amp.localhost:8080/oauth2/token
 IDP_CLIENT_ID=amp-api-client
 IDP_CLIENT_SECRET=amp-api-client-secret
 
+# Required by security/roles to create and remove disposable role personas.
+# Override all three with secret-backed values outside local quick-start.
+THUNDER_ADMIN_URL=http://thunder.amp.localhost:8080
+# Optional when Thunder's System resource identifier is not <admin-url>/mcp.
+# THUNDER_SYSTEM_RESOURCE=https://thunder.example.com/mcp
+THUNDER_SYSTEM_CLIENT_ID=amp-system-client
+THUNDER_SYSTEM_CLIENT_SECRET=amp-system-client-secret
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/.env.example:20]: Ignored because these credentials used only by the local development environment not a real secret

File: test/e2e/fixtures/security-probe-agent/probe/app.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+@app.post("/security/identity")
+async def identity() -> dict[str, object]:
+    return (await mint_agent_token()).public()
+
+
+@app.post("/security/mcp/{tool}")
+async def mcp(tool: str) -> dict[str, object]:
+    if tool not in TOOL_ARGUMENTS:
+        raise HTTPException(status_code=404, detail="unknown fixed tool")
+    return await probe_mcp_tool(tool)
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/fixtures/security-probe-agent/probe/app.py:39]: Ignored because container is not directly exposed publicly, and its operations are restricted to fixed, harmless targets. It never returns access tokens or credentials

File: test/e2e/fixtures/security-probe-agent/probe/mcp.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    if not proxy_url:
+        return {
+            "tool": tool,
+            "phase": "configuration",
+            "token_minted": False,
+            "http_status": None,
+            "authorized": False,
+            "result_received": False,
+            "error": "mcp_url_not_configured",
+        }
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/fixtures/security-probe-agent/probe/mcp.py:41]: Ignored because the missing value response is therefore an intentional test state, not a startup configuration error.

File: test/e2e/fixtures/security-probe-agent/probe/runtime.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+def runtime_posture() -> dict[str, object]:
+    """Return booleans describing the sandbox without exposing its contents."""
+
+    root_filesystem_read_only = False
+    root_probe = Path("/security-probe-root-write")
+    try:
+        root_probe.write_text("probe", encoding="utf-8")
+        root_probe.unlink(missing_ok=True)
+    except OSError:
+        root_filesystem_read_only = True
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/fixtures/security-probe-agent/probe/runtime.py:34]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: test/e2e/fixtures/security-probe-agent/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+fastapi==0.116.1
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/fixtures/security-probe-agent/requirements.txt]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: test/e2e/framework/persona.go [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+	appID, clientID, clientSecret, err := p.createApplication(ctx, ouID, name)
+	if err != nil {
+		return nil, err
+	}
+	client := &AudienceClient{
+		Name:         name,
+		appID:        appID,
+		clientID:     clientID,
+		clientSecret: clientSecret,
+	}
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/framework/persona.go:194]: Ignored. This test specifically needs the generated OAuth client ID/audience to start with amp-publisher-. Adding e2e-test- would change the audience and invalidate the publisher-boundary test. The suite already registers cleanup immediately after creating these clients.

File: test/e2e/security/authz/scope_matrix_test.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		})
+
+		It(fmt.Sprintf("allows %s to a token with %s",
+			route.Method+" "+route.Path, route.Scopes[0]), func() {
+			By("calling the same route with only the scope it requires")
+			allow := clientWithScopes([]string{route.Scopes[0]})
+			resp := send(allow, route)
+			defer resp.Body.Close()
+			framework.ExpectNotForbidden(Default, resp, route.String())
+		})
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/authz/scope_matrix_test.go:163]: Ignored. ExpectNotForbidden already rejects both 401 and 403. It intentionally permits downstream 5xx responses because the authorization check has been passed, and the team agreed these known handler errors should be recorded without failing the security suite ATM.

File: test/e2e/security/authz/scope_matrix_test.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	Expect(err).NotTo(HaveOccurred(), "failed to decode the issued token")
+	for _, s := range framework.AllScopes() {
+		if !slices.Contains(scopes, s) {
+			Expect(issued).NotTo(HaveKey(s),
+				"the IDP issued %q which was not requested — this spec would be vacuous", s)
+		}
+	}
+
+	c := framework.NewAMPClientWithToken(Cfg, token)
+	tokenCache[key] = c
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/authz/scope_matrix_test.go]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: test/e2e/security/observability/suite_test.go [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+// verifyObserverReachable fails fast with an actionable message rather than
+// letting every spec fail on a connection error. The observer is a separate
+// deployment from agent-manager-service, so it can be absent or unexposed on a
+// cluster where the rest of the platform is fine.
+func verifyObserverReachable() {
+	resp, err := http.Get(obsURL("/api/v1/traces", nil)) //nolint:gosec // fixed test-config URL
+	Expect(err).NotTo(HaveOccurred(),
+		"ABORTING: cannot reach the observer at %s. Set AM_OBSERVER_BASE_URL, or check the "+
+			"observability extension is installed and port-forwarded.", Cfg.ObserverBaseURL)
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/observability/suite_test.go]: fixed in [fb84de9](https://github.com/wso2/agent-manager/pull/1676/commits/fb84de9cdba10030bc6e8320352326527136b252)

File: test/e2e/security/README.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+The vulnerable implementation embedded the git secret's username and password
+in the clone URL, causing `git clone` to persist them in
+`/mnt/vol/source/.git/config`. The build step mounts that same workspace, so a
+user-supplied Dockerfile could read them across the `git-secret:read` boundary.
+The fixed checkout workflow supplies the credential from a checkout-container
+file under `/tmp` and leaves the recorded remote URL credential-free.
+
+Add a permanent regression test proving the credential is absent from the
+build workspace, build logs, image layers, and final image.
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/README.md]: Removed outdated comments in [ef967c8](https://github.com/wso2/agent-manager/pull/1676/commits/ef967c8a736dcdedbe2157b5e752b4ee2cff79b4)

File: test/e2e/security/README.md [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+the day multi-org lands, a handler reading the path becomes a tenancy breach on
+arrival.
+
+Two known items to revisit when multi-org work starts:
+
+- `agent-manager-observer` takes `organization` from the query string and never
+  compares it to the token. Inert today (`NamespaceFor` ignores the argument and
+  returns a single namespace), a cross-tenant read the moment it does not.
+- Collection and search endpoints need tenant-filtering coverage; per-resource
+  GETs are the easy half.
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/README.md]: Removed outdated comments in [ef967c8](https://github.com/wso2/agent-manager/pull/1676/commits/ef967c8a736dcdedbe2157b5e752b4ee2cff79b4)

File: test/e2e/security/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/security/README.md]: The underlying Git credential leak has already been fixed on main. This PR introduces the initial security conformance suite and does not claim complete security coverage. Testing build workspaces and image layers requires additional build-pipeline inspection support that is outside this PR’s scope, so the regression test will be tracked separately.
https://github.com/wso2/agent-manager/commit/7cbfdfbe89e23cd8d5a886d19311f358c1d04055

File: test/e2e/framework/client.go [hunk 15]
```diff
@@ -62,6 +63,11 @@ func (c *AMPClient) Token() string {
 // Do sends an HTTP request to the AMP API. If body is non-nil it is marshaled to JSON.
 // The path is appended to the base URL (e.g., "/api/v1/orgs").
 func (c *AMPClient) Do(method, path string, body any) (*http.Response, error) {
+	return c.DoWithContext(context.Background(), method, path, body)
```
[the remark this answers is not in the document]
Comment (vish-mv) [reply test/e2e/framework/client.go:72]: Addressed for the security suite: all security-test HTTP operations now propagate SpecContext, and PatchWithContext was added. The context-free wrappers are retained for backward compatibility with existing E2E callers. Migrating the entire E2E suite is outside this PR’s scope and should be handled separately.

**Review by menakaj — APPROVED** [review]
